ProjectUpdater/src/au/gov/asd/tac/projectupdater/ProjectUpdater.java

```diff
@@ -0,0 +1,267 @@
+                    Node nextNode = classPathExtensionNode.getNextSibling();
+                    while (nextNode instanceof Text) {
+                        Node textNode = nextNode;
+                        nextNode = nextNode.getNextSibling();
+                        dataNode.removeChild(textNode);
+                    }
+                    Node prevNode = classPathExtensionNode.getPreviousSibling();
+                    while (prevNode instanceof Text) {
+                        Node textNode = prevNode;
+                        prevNode = prevNode.getPreviousSibling();
+                        dataNode.removeChild(textNode);
+                    }
+                    dataNode.removeChild(classPathExtensionNode);
+                } else {
+                    throw new IllegalStateException("Not a valid project.xml file");
+                }
+            }
+
+            // Restore each jar file to the project.xml
+            // If the jar directory does not exist then assume no wrapped jars
+            // Ensure each file is a JAR file
+            // Add white space to make it look the same as the Netbeans version
+            if (jarDirectory.exists()) {
+                for (File jarFile : jarDirectory.listFiles()) {
+                    if (jarFile.getName().endsWith(".jar")) {
+                        logMessage("\tIncluding jar file: " + jarFile.getName());
+                        addClassPathExtension(document, jarFile, dataNode);
+
+                        if (publicPackagesFile.exists()) {
+                            extractMatchingPackages(jarFile, packages, publicPackages);
+                        }
+                    } else {
+                        throw new IllegalStateException("Not a JAR file: " + jarFile.getAbsolutePath());
+                    }
+                }
+
+                dataNode.appendChild(document.createTextNode("\n        "));
+            }
+
+            if (publicPackagesFile.exists()) {
+                for (String publicPackage : publicPackages) {
+                    if (!publicPackage.startsWith("META-INF")) {
+                        Element publicPackageElement = document.createElement("package");
+                        publicPackageElement.setTextContent(publicPackage);
+                        publicPackagesNode.appendChild(document.createTextNode("\n                "));
+                        publicPackagesNode.appendChild(publicPackageElement);
+                    }
+                }
+            }
+            if (publicPackagesNode != null) {
+                publicPackagesNode.appendChild(document.createTextNode("\n            "));
+            }
+
+            // Delete the existing old file and replace it with a copy of the current project.xml
+            oldProjectFile.delete();
+            projectFile.renameTo(oldProjectFile);
+
+            // Save the edited document to project.xml
+            saveXMLFile(document, projectFile);
+        } catch (Exception ex) {
```

Comment:
Is there a more specific set of exceptions that can be caught in this instance? (Can't tell just glancing at the code whether the generic exception needs to be caught or not)

Reply:
The way this is written the generic exception works out the best (because there are too many possible exceptions). It will also pick up run time errors this way. Cool?